ota_proxy/ota_cache.py

```diff
@@ -0,0 +1,453 @@
+            hash_list, files_list = [], []
+            space_available = None
+            for h in self:
+                if space_available >= size:
+                    break
+                else:
+                    f: Path = Path(cfg.BASE_DIR) / h
+                    if f.is_file():
+                        # TODO: deal with dangling cache?
+                        space_available += f.stat().st_size
+                        hash_list.append(h)
+                        files_list.append(f)
+
+            if space_available >= size:
+                enough_space = True
+                # we can reserve enough space from current bucket
+                for h in hash_list:
+                    del self[h]
+
+        if enough_space:
+            for f in files_list:
+                f.unlink(missing_ok=True)
+            return hash_list
+
+        return
+
+
+class OTAFile:
+    def __init__(
+        self,
+        url: str,
+        fp: typing.BinaryIO,
+        meta: db.CacheMeta,
+        store_cache: bool = False,
+        free_space_event: Event = None,
+    ):
+        logger.debug(f"new OTAFile request: {url}")
+        self._store_cache = store_cache
+        self._free_space_event = free_space_event
+        # NOTE: for new cache entry meta, the hash and size are not set yet
+        self.meta = meta
+
+        # data stream
+        self._fp = fp
+        self._queue = None
+
+        # life cycle
+        self.finished = False
+        self.cached_success = False
+
+        # prepare for data streaming
+        if store_cache:
+            self._hash_f = sha256()
+            self._queue = Queue()
+
+    def background_write_cache(self, callback):
+        if not self._store_cache or not self._free_space_event:
+            return
+
+        try:
+            logger.debug(f"start to cache for {self.meta.url}...")
+            tmp_fpath = Path(cfg.BASE_DIR) / str(time.time()).replace(".", "")
+            self.temp_fpath = tmp_fpath
+            self._dst_fp = open(tmp_fpath, "wb")
+
+            while not self.finished or not self._queue.empty():
+                if not self._free_space_event.is_set():
+                    # if the free space is not enough duing caching
+                    # abort the caching
+                    logger.error(
+                        f"not enough free space during caching url={self.meta.url}, abort"
+                    )
+                    break
+
+                try:
+                    data = self._queue.get(timeout=360)
+                except Exception:
+                    logger.error(f"timeout caching for {self.meta.url}, abort")
+                    break
+
+                self._hash_f.update(data)
+                self.meta.size += self._dst_fp.write(data)
+
+            self._dst_fp.close()
+            if self.finished and self.meta.size > 0:  # not caching 0 size file
+                # rename the file to the hash value
+                hash = self._hash_f.hexdigest()[:16]
```

Comment:
whole hash value should be used. sha256 is unique in 256 bits.
```suggestion
                hash = self._hash_f.hexdigest()
```